### PR TITLE
ENT-4063-preserve_during_upgrade.txt PR

### PR DESCRIPTION
**Pull request by Lex-2008**

### CODE REVIEW

**Comment by Lex-2008** [comment]
@cf-bottom build, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1197/)

https://ci.cfengine.com/job/pr-pipeline/1197/

**Comment by Lex-2008** [comment]
thankyou, @cf-bottom

**Comment by cf-bottom** [comment]
I'm not sure I understand, @Lex-2008.

File: packaging/common/cfengine-hub/preinstall.sh [hunk 1]
```diff
@@ -200,20 +200,30 @@ if [ -d $PREFIX/httpd/htdocs ]; then
...
+  if [ -f $PREFIX/httpd/htdocs/preserve_during_upgrade.txt ]; then
+    # Purge all files in httpd/htdocs with exceptions listed in preserve_during_upgrade.txt
+    cf_console echo "Keeping only what's listed in preserve_during_upgrade.txt file"
+    $PRESERVE_FILTER="`sed -e '
+      /^\s*#/d  # skip lines beginning with #
+      /^\s*$/d  # also skip empty lines
+      \_/$_{    # for lines ending with /
+                # treat them as dirnames and exclude all files inside:
+        s_.*_-not \( -path "PREFIX/httpd/htdocs/&" -prune \)_
+      }
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/preinstall.sh]: Please put this filter into a function with a documentation of its behavior.

File: packaging/common/cfengine-hub/preinstall.sh [hunk 2]
```diff
@@ -200,20 +200,30 @@ if [ -d $PREFIX/httpd/htdocs ]; then
   tar zcf /tmp/cfengine-htdocs.tar.gz $PREFIX/httpd/htdocs
 
   cf_console echo "Purging old version from $PREFIX/httpd/htdocs"
-  if [ -d $PREFIX/share/GUI ]; then
+  if [ -f $PREFIX/httpd/htdocs/preserve_during_upgrade.txt ]; then
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/preinstall.sh]: So who/what is responsible for creating this file? I recall it was in some @aleksandrychev 's commit? Could you please reference it in the commit message?

Comment (Lex-2008) [reply packaging/common/cfengine-hub/preinstall.sh]: done!

**Comment by aleksandrychev** [comment]
@Lex-2008 We need also this tooling for 3.10 and 3.12

**Pull request merged**